zipkin-collector/core/src/main/java/zipkin2/collector/Collector.java

```diff
@@ -110,34 +119,46 @@ public void accept(List<Span> spans, Callback<Void> callback) {
     }
 
     try {
+      // adding a separate callback intentionally decouples collection from storage
       record(sampled, acceptSpansCallback(sampled));
       callback.onSuccess(null);
-    } catch (RuntimeException e) {
-      callback.onError(errorStoringSpans(sampled, e));
-      return;
+    } catch (RuntimeException | Error e) {
+      handleStorageError(spans, e, callback);
```

Comment:
Technically speaking, isn't this still a "Collector error" at this point?  Since `record()` is async, the "storage error" doesn't happen until later.  Arguably, we may want to handle them the same way but I wanted to bring up the potential issue just in case.

Reply:
thanks for asking. so the most likely cause at this point is in the "record" method, which I suppose we should rename to store. There's always a chance that there's an error in the storage implementation prior to returning a call. For example, in Elasticsearch a failure to install the index template would propagate here.

```
  void record(List<Span> sampled, Callback<Void> callback) {
    storage.spanConsumer().accept(sampled).enqueue(callback);
  }
```

Reply:
Interesting point!  Would it be worthwhile to separate the `...accept()` call from `enqueue()` for the sake of error messaging?  I'm not exactly sold on the difference being worthwhile at the moment.

Reply:
added a commit on this topic. good question

Reply:
sorry wires crossed. So, In any case I think we'll still need to invoke either callback.onSuccess or onError.

So I guess you are asking if the code added to separate out which of the failures into separate log prefixes would be worth it..

ex instead of just "storage error" a separate if statement and log message for `storage.spanConsumer()` vs `spanConsumer.accept()` vs `call.enqueue()`.

I can understand that in the logs you might not have the stack trace (though the caller may), so this could be a way to know the difference, if the message wasn't already clear what happened..

Reply:
sorry it would be like instead of

"Cannot store spans ... due to Error installing ES template" (or whatever the cause would be)
"Cannot get instance of span consumer ...
or
"Cannot create storage call to accept spans ...
or
"Cannot enqueue spans for storage...

I think classifying like this would add a dozen lines of main and 3 or 4 more test cases, but thinking right now, I'm not sure it would be terribly more helpful.